src/metabase/actions/execution.clj

```diff
@@ -0,0 +1,107 @@
+(ns metabase.actions.execution
+  (:require
+    [clojure.tools.logging :as log]
+    [medley.core :as m]
+    [metabase.actions.http-action :as http-action]
+    [metabase.api.common :as api]
+    [metabase.models :refer [Dashboard DashboardCard]]
+    [metabase.models.action :as action]
+    [metabase.query-processor.error-type :as qp.error-type]
+    [metabase.query-processor.writeback :as qp.writeback]
+    [metabase.util :as u]
+    [metabase.util.i18n :refer [tru]]
+    [toucan.db :as db]))
+
+(defn- map-parameters
+  "Take the `parameters` map passed in to an endpoint and map it to the parameters
+  in the underlying `Action` so they can be attached to the query that gets passed to
+   [[qp/execute-write-query!]] or [[execute-http-action!]].
+
+  Incoming API request `:parameters` should look like
+
+    [{:id \"my_id\",
+      :type \"id\",
+      :value \"12\"}]
+
+  And `parameter_mappings` should look like
+
+     [{:parameter_id \"my_id\", :target [:variable [:template-tag \"id\"]]}]
+
+  We need to convert these to a list like
+
+    [{:target [:variable [:template-tag \"id\"]]
+      :value  \"12\"}]
+
+  before passing to the QP code."
+  [parameters parameter-mappings]
+  (let [params-by-id (m/index-by :id parameters)
+        mappings-by-id (m/index-by :parameter_id parameter-mappings)]
+    (mapv (fn [[param-id param-value-info]]
+            (let [target (or (get-in mappings-by-id [param-id :target])
+                             (throw (ex-info (tru "No parameter mapping found for parameter {0}. Found: {1}"
```

Comment:
I think it would be more precise but less helpful, as the mapping is really the part that gets set up by a user, and the param code will bring the target in.